### PR TITLE
Ensure that no partial toolchain is left behind

### LINKED ISSUE #14761: Canceled toolchain download leaves broken download behind
If one cancels the build while a toolchain is downloading, we leave a partial download behind. The next build will pick this up (assuming its a fully downloaded toolchain) and try to provision it (but fails due to it being a broken file).

### PR DESCRIPTION
**Pull request by bmuskalla**
Fixes #14761

### CODE REVIEW

**Comment by bmuskalla** [comment]
Ignore the first 2 commits (will be gone after rebase)

**Comment by bmuskalla** [comment]
@bot-gradle test QF

File: subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/install/internal/AdoptOpenJdkDownloader.java [hunk 1]
```diff
@@ -69,11 +69,17 @@ public String getShortDisplayName() {
...
-            copyIntoFile(source, inputStream, tmpFile);
-        });
+    private void downloadResource(URI source, File targetFile, ExternalResource resource) {
+        final File downloadFile = new File(targetFile.getAbsoluteFile() + ".part");
+        try {
+            resource.withContent(inputStream -> {
+                LOGGER.info("Downloading {} to {}", resource.getDisplayName(), targetFile);
+                copyIntoFile(source, inputStream, downloadFile);
+            });
+            downloadFile.renameTo(targetFile);
```
Comment (ljacomet) [inline subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/install/internal/AdoptOpenJdkDownloader.java:79]: We should not ignore the return value here.
I also believe that using `java.nio.file.Files#move` with the right options might be a better choice.

**Review by ljacomet — CHANGES_REQUESTED** [review]
See comment

**Comment by bmuskalla** [comment]
@bot-gradle test this

**Review by ljacomet — APPROVED** [review]

**Pull request merged**